File: ch_item_master/ch_item_master/report/items_without_active_price/items_without_active_price.py
# Copyright (c) 2026, GoStack and contributors
# For license information, please see license.txt

"""
Items Without Active Price — Script Report

Shows all CH-managed items (variant items with ch_model set) that have NO
Active price record in any channel. Helps ensure every sellable item has pricing.
"""

import frappe
from frappe import _


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {"fieldname": "item_code", "label": _("Item Code"), "fieldtype": "Link",
         "options": "Item", "width": 150},
        {"fieldname": "item_name", "label": _("Item Name"), "fieldtype": "Data", "width": 250},
        {"fieldname": "ch_model", "label": _("Model"), "fieldtype": "Link",
         "options": "CH Model", "width": 150},
        {"fieldname": "ch_sub_category", "label": _("Sub Category"), "fieldtype": "Link",
         "options": "CH Sub Category", "width": 150},
        {"fieldname": "ch_category", "label": _("Category"), "fieldtype": "Link",
         "options": "CH Category", "width": 120},
        {"fieldname": "last_expired", "label": _("Last Expired Price"), "fieldtype": "Date", "width": 130},
        {"fieldname": "total_prices", "label": _("Total Prices (all time)"), "fieldtype": "Int", "width": 130},
    ]


def get_data(filters):
    conditions = ["i.ch_model IS NOT NULL", "i.ch_model != ''",
                   "i.has_variants = 0", "i.disabled = 0"]
    values = {}

    if filters.get("category"):
        conditions.append("i.ch_category = %(category)s")
        values["category"] = filters["category"]

    if filters.get("sub_category"):
        conditions.append("i.ch_sub_category = %(sub_category)s")
        values["sub_category"] = filters["sub_category"]

    if filters.get("model"):
        conditions.append("i.ch_model = %(model)s")
        values["model"] = filters["model"]

    return frappe.db.sql("""
        SELECT
            i.item_code,
            i.item_name,
            i.ch_model,
            i.ch_sub_category,
            i.ch_category,
            (
                SELECT MAX(p.effective_to)
                FROM `tabCH Item Price` p
                WHERE p.item_code = i.item_code AND p.status = 'Expired'
            ) AS last_expired,
            (
                SELECT COUNT(*)
                FROM `tabCH Item Price` p2
                WHERE p2.item_code = i.item_code
            ) AS total_prices
        FROM `tabItem` i
        WHERE {conditions}
          AND NOT EXISTS (
            SELECT 1 FROM `tabCH Item Price` ap
            WHERE ap.item_code = i.item_code AND ap.status = 'Active'
          )
        ORDER BY i.ch_category, i.ch_sub_category, i.item_name
    """.format(conditions=" AND ".join(conditions)), values, as_dict=True)

File: ch_item_master/ch_item_master/report/items_without_active_price/items_without_active_price.js
// Copyright (c) 2026, GoStack and contributors
// Items Without Active Price — Report filters

frappe.query_reports["Items Without Active Price"] = {
	filters: [
		{
			fieldname: "category",
			label: __("Category"),
			fieldtype: "Link",
			options: "CH Category",
			on_change: (query_report) => {
				// Cascade: clear sub category / model if they no longer belong to the chosen category
				const cat = query_report.get_filter_value("category");
				const sub = query_report.get_filter_value("sub_category");
				if (cat && sub) {
					frappe.db.get_value("CH Sub Category", sub, "category").then((r) => {
						if (((r.message || {}).category || "") !== cat) {
							query_report.set_filter_value({ sub_category: "", model: "" });
						} else {
							query_report.refresh(true);
						}
					});
				} else {
					query_report.refresh(true);
				}
			},
		},
		{
			fieldname: "sub_category",
			label: __("Sub Category"),
			fieldtype: "Link",
			options: "CH Sub Category",
			get_query() {
				let cat = frappe.query_report.get_filter_value("category");
				return cat ? { filters: { category: cat } } : {};
			},
			on_change: (query_report) => {
				if (query_report._no_refresh) return;
				// Cascade: clear the model if it no longer belongs to the chosen sub category
				const sub = query_report.get_filter_value("sub_category");
				const model = query_report.get_filter_value("model");
				if (sub && model) {
					frappe.db.get_value("CH Model", model, "sub_category").then((r) => {
						if (((r.message || {}).sub_category || "") !== sub) {
							query_report.set_filter_value("model", "");
						} else {
							query_report.refresh(true);
						}
					});
				} else {
					query_report.refresh(true);
				}
			},
		},
		{
			fieldname: "model",
			label: __("Model"),
			fieldtype: "Link",
			options: "CH Model",
			get_query() {
				const filters = {};
				const sc = frappe.query_report.get_filter_value("sub_category");
				const cat = frappe.query_report.get_filter_value("category");
				if (sc) filters.sub_category = sc;
				else if (cat) filters.category = cat;
				return {
					query: "ch_item_master.ch_item_master.api.search_ch_models",
					filters: filters,
				};
			},
		},
	],
};
